lib/alg/extract-path.js

```diff
@@ -0,0 +1,28 @@
+module.exports = extractPath;
+
+function extractPath(shortestPaths, source, destination) {
+  if (shortestPaths[source].predecessor !== undefined) {
+    throw new Error("Invalid source vertex");
+  }
+
+  if (shortestPaths[destination].predecessor === undefined  && destination !== source) {
+    throw new Error("Invalid destination vertex");
+  }
+
+  return {
+    weight: shortestPaths[destination].distance,
+    path: runExtractPath(shortestPaths, source, destination)
+  };
+}
+
+function runExtractPath(shortestPaths, source, destination) {
+  if (destination === source) {
+    return [source];
+  }
+
+  return runExtractPath(
```

Comment:
I'm pretty concerned about the recursion here. Are you sure that it works for paths with a big length? Check if it works, and if not rewrite it using iterations.

Reply:
thanks for the feedback @pkakelas 
I wrote some c++ code to generate test js files:
```
#include <cstdio>

using namespace std;

const int maxLength = 8372;

int main() {
  printf("var Graph = require('./graphlib').Graph;\nvar alg = require('./graphlib').alg;\nvar g = new Graph();\n");
  printf("var shortestPaths={\n\"0\": { distance: 0 },\n");
  for(int i = 1; i <= maxLength; i++){
    printf("\"%d\": { distance: 73, predecessor: \"%d\" },\n",i,i-1);
  }
  printf("};\n");
  printf("console.log(alg.extractPath(shortestPaths,\"0\",\"%d\"));",maxLength);
}

```
And by binary searching on the length of the path, I found that it reaches maximum recursion depth at ~8400. 
Your concerns were right :p I will rebuild this using iterations

Reply:
Good catch.

Reply:
@mstou Can you apply the same stress test to #93? It seems that both the old and the new implementation of `dfs` may be susceptible to stack overflows. In that case I will refractor it to be iterative. It seems that an example of a chain graph is appropriate for this stress test.

Reply:
okay @dionyziz im on it

Reply:
Thanks!